### ISSUE TITLE
Go Text Protocol

### ISSUE THREAD

**Issue by wrongu** [Rochester-NRT/RocAlphaGo#52]
[GTP](http://senseis.xmp.net/?GoTextProtocol) appears to be the standard protocol for computer go programs. Should implement to/from GTP functions (see #51)

**Comment by jtauber** [comment]
I'm interested in implementing GTP alongside my existing SGF library.

**Comment by Nefeste** [comment]
https://github.com/pasky/pachi/blob/a8e492c7e387fb1443881ca4a695eaabeefd706b/gtp.c
https://github.com/pasky/pachi/blob/a8e492c7e387fb1443881ca4a695eaabeefd706b/gtp.h
this could be useful probably..

**Comment by wrongu** [comment]
@jtauber that would be great! I'm curious how you see it interacting with your SGF library - or am I misunderstanding?

[Here](https://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html) are the GTP specs I am looking at. We'd need the 'engine' not the 'controller', and an instantiation of an engine should have a reference to a particular player. In our code, a `player` will expect a `GameState` object as input. Basically, I imagine a GTP engine that simply wraps a player + state - something like

```python
class GTPEngine(object):
    def __init__(self, player, size=19):
        self.board_state = GameState(size)
        self.player = player
    def listen(self, addr, port):
        # start listening for server commands on given address/port
        # when a message is received, call the relevant handler
    def msg_clear_board(self):
        # for example
        self.board_state = GameState(self.board_state.size)
    def msg_genmove(self, color):
        move = self.player.get_move(self.state)
        # now send move response back to controller...
    # ...
    # ...
```

**Comment by jtauber** [comment]
I'm working on just parsing the messages and generating the responses initially. As of last night, I was half way through the "required" commands.

My SGF library just converts to and from the SGF file format and a Python data structure representing collections, games and variations. I can imagine a script built on both my SGF and GTP libraries to do things like "replay" an SGF game as a sequence of GTP commands or persist a sequence of GTP commands as an SGF game.

The third key component to all this is the game state so when I get far enough along (next few days) I'll work out a way of making the game state pluggable.

**Comment by amaso13** [comment]
Hi everyone,
I just joined GitHub and this group. I am a doctoral student in Information Systems, Analytics & Decision Support. I am new in Data Science, can anyone help with how to get started and how to contribute on this?
Thanks for your orientation.

Armel.

**Comment by jtauber** [comment]
As well as working through the GTP spec I'm testing behavior against GnuGo just to make sure I'm reading the spec correctly.

**Comment by jtauber** [comment]
You can now follow along at https://github.com/jtauber/gtp

All the required GTP commands are implemented from a parsing point of view. What remains is the networking stuff and making the game state / player pluggable so it can be integrated into this project.

Neither of those should take long.

**Comment by wrongu** [comment]
All sounds good! Since the GTP engine is responsible for knowing state/move legality/etc, it made sense to me to use the `GameState` we've defined here. The alternative - writing state conversion functions from our state to your state and back - is not crazy either especially if you want your project to be a standalone python GTP library.

That said, if you have in mind a way that the state itself could be 'pluggable' that would be the best of both worlds!

**Comment by jtauber** [comment]
Now that my SGF library is in pretty good shape (as best I can tell, I'll shift my focus back to GTP library).

**Comment by jtauber** [comment]
@wrongu I've merged your PRs (with one fix) and done a 0.2 release.

Can this issue  be closed now?
